crates/futures/src/task/multithread.rs

```diff
@@ -0,0 +1,201 @@
+            return;
+        }
+
+        debug_assert_eq!(prev, SLEEPING);
+
+        notify(&this.state, 1);
+    }
+
+    // This is to avoid a dependency on futures_util::task::ArcWake
+    unsafe fn into_raw_waker(this: Arc<Self>) -> RawWaker {
+        unsafe fn raw_clone(ptr: *const ()) -> RawWaker {
+            let ptr = ManuallyDrop::new(Arc::from_raw(ptr as *const AtomicWaker));
+            AtomicWaker::into_raw_waker((*ptr).clone())
+        }
+
+        unsafe fn raw_wake(ptr: *const ()) {
+            let ptr = Arc::from_raw(ptr as *const AtomicWaker);
+            AtomicWaker::wake_by_ref(&ptr);
+        }
+
+        unsafe fn raw_wake_by_ref(ptr: *const ()) {
+            let ptr = ManuallyDrop::new(Arc::from_raw(ptr as *const AtomicWaker));
+            AtomicWaker::wake_by_ref(&ptr);
+        }
+
+        unsafe fn raw_drop(ptr: *const ()) {
+            drop(Arc::from_raw(ptr as *const AtomicWaker));
+        }
+
+        const VTABLE: RawWakerVTable =
+            RawWakerVTable::new(raw_clone, raw_wake, raw_wake_by_ref, raw_drop);
+
+        RawWaker::new(Arc::into_raw(this) as *const (), &VTABLE)
+    }
+}
+
+
+thread_local! {
+    static QUEUE: Queue<Rc<Task>> = Queue::new();
+}
+
+
+struct Inner {
+    future: Pin<Box<dyn Future<Output = ()> + 'static>>,
+    closure: Closure<dyn FnMut(JsValue)>,
+}
+
+pub(crate) struct Task {
+    atomic: Arc<AtomicWaker>,
+    waker: Waker,
+    // This is an Option so that the Future and Closure can be immediately dropped when it's finished
+    inner: RefCell<Option<Inner>>,
+}
+
+impl Task {
+    pub(crate) fn spawn(future: Pin<Box<dyn Future<Output = ()> + 'static>>) {
+        let atomic = AtomicWaker::new();
+
+        let waker = unsafe { Waker::from_raw(AtomicWaker::into_raw_waker(Arc::clone(&atomic))) };
+
+        let this = Rc::new(Task {
+            atomic,
+            waker,
+            inner: RefCell::new(None),
+        });
+
+        let closure = {
+            let this = Rc::clone(&this);
+
+            Closure::wrap(Box::new(move |_| Task::run(&this)) as Box<dyn FnMut(JsValue)>)
+        };
+
+        *this.inner.borrow_mut() = Some(Inner { future, closure });
+
+        QUEUE.with(move |queue| {
+            // This will cause it to call crate::queue::Task::run on the next microtask tick
+            queue.push_task(this);
+        });
+    }
+
+    fn run(&self) {
+        let mut borrow = self.inner.borrow_mut();
+
+        // This will only be None if the Future wakes up the Waker after returning Poll::Ready
+        if let Some(inner) = borrow.as_mut() {
+            // This is so that it will re-queue if poll calls waker.wake()
+            self.atomic.sleep();
+
+            let poll = {
+                let cx = &mut Context::from_waker(&self.waker);
+                inner.future.as_mut().poll(cx)
+            };
+
+            if let Poll::Ready(_) = poll {
+                // This prevents it from notifying again
+                self.atomic.done();
```

Comment:
I think this is fine to omit, it's saving at most one `atomic.notify` instruction and everything else here is so expensive I think it's safe to just omit this and deal with it later if necessary.